crates/accelerate/src/sabre/route.rs

```diff
@@ -0,0 +1,584 @@
+                        {
+                            // 2Q op that cannot be placed. Add it to the front layer and move on.
+                            self.front_layer.insert(
+                                node_id,
+                                [a.to_phys(&self.layout), b.to_phys(&self.layout)],
+                            );
+                            continue;
+                        }
+                        _ => {}
+                    }
+                }
+            }
+
+            // If we reach here, the node is routable.
+            self.gate_order.push(node.py_node_id);
+            for edge in dag.dag.edges_directed(node_id, Direction::Outgoing) {
+                let successor_node = edge.target();
+                let successor_index = successor_node.index();
+                self.required_predecessors[successor_index] -= 1;
+                if self.required_predecessors[successor_index] == 0 {
+                    to_visit.push(successor_node);
+                }
+            }
+        }
+    }
+
+    /// Inner worker to route a control-flow block.  Since control-flow blocks are routed to
+    /// restore the layout at the end of themselves, and the recursive calls spawn their own
+    /// tracking states, this does not affect our own state.
+    fn route_control_flow_block(&self, block: &SabreDAG) -> BlockResult {
+        let (result, mut block_final_layout) =
+            swap_map_trial(self.target, block, self.heuristic, &self.layout, self.seed);
+        // For now, we always append a swap circuit that gets the inner block back to the
+        // parent's layout.
+        let swap_epilogue = {
+            // Map physical location in the final layout from the inner routing to the current
+            // location in the outer routing.
+            let mapping: HashMap<NodeIndex, NodeIndex> = block_final_layout
+                .iter_physical()
+                .map(|(p, v)| {
+                    (
+                        NodeIndex::new(p.index()),
+                        NodeIndex::new(v.to_phys(&self.layout).index()),
+                    )
+                })
+                .collect();
+
+            let swaps = token_swapper(
+                &self.target.coupling,
+                mapping,
+                Some(SWAP_EPILOGUE_TRIALS),
+                Some(self.seed),
+                None,
+            )
+            .unwrap();
+
+            // Convert physical swaps to virtual swaps
+            swaps
+                .into_iter()
+                .map(|(l, r)| {
+                    let p_l = PhysicalQubit::new(l.index().try_into().unwrap());
+                    let p_r = PhysicalQubit::new(r.index().try_into().unwrap());
+                    block_final_layout.swap_physical(p_l, p_r);
+                    [p_l, p_r]
+                })
+                .collect()
+        };
+        BlockResult {
+            result,
+            swap_epilogue,
+        }
+    }
```

Comment:
Since this is now a separate call, I inlined `gen_swap_epilogue` - it's the majority of the logic already, so further abstraction wouldn't have helped much.